packages/admin-panel/src/VizBuilderApp/views/Main.js

```diff
@@ -21,9 +26,50 @@ const RightCol = styled(FlexColumn)`
   flex: 1;
 `;
 
+const StyledAlert = styled(SmallAlert)`
+  margin-top: 30px;
+  margin: auto;
+`;
+
+const Progress = styled(CircularProgress)`
+  margin-top: 30px;
+  margin: auto;
+`;
+
 // Todo: add warning on page unload https://github.com/jacobbuck/react-beforeunload#readme
-export const Main = () => {
+export const Main = ({ match }) => {
+  const { visualisationId } = match.params;
+
+  // do not fetch existing viz if no visualisationId is provided in the params
+  const fetchExistingVizEnabled = visualisationId !== undefined;
+
+  const [_, { setVisualisation }] = useVizBuilderConfig();
   const [enabled, setEnabled] = useState(false);
+  const [visualisationLoaded, setVisualisationLoaded] = useState(false);
+  const { data = {}, error } = useDashboardVisualisation(visualisationId, fetchExistingVizEnabled);
+  const { visualisation } = data;
+
+  useEffect(() => {
+    if (visualisation) {
+      setVisualisation(visualisation);
+      setVisualisationLoaded(true);
+    }
+  }, [visualisation]);
+
+  // Show error if failed to load an existing visualisation
+  if (visualisationId && error) {
+    return (
+      <StyledAlert severity="error" variant="standard">
+        {error.message}
+      </StyledAlert>
+    );
+  }
+
+  // Wait until visualisation is loaded to populated the field correctly if we are viewing an existing viz
+  if (visualisationId && !visualisationLoaded) {
+    return <Progress size={100} />;
+  }
```

Comment:
There's a bit of a problem with JSON Edit box that it doesn't automatically re-render the json value inside the box if the value is changed. So I thought it'd be better to wait until the viz is done fetched and then render the whole component. Also if there's any error, don't render anything

Reply:
That seems a bit strange since it is a controlled component and the state is kept externally the whole time. Hopefully it doesn't create other issues not updating. Anyway this looks like a good workaround and probably removes a whole lot of edge cases rendering things before the data is fetched 👍